database/migrations/default/1742417174000_crash_ems_link/up.sql

```diff
@@ -0,0 +1,391 @@
+        when
+            lower(pcr_cause_of_injury) like '%bike%'
+            or lower(pcr_cause_of_injury) like '%bicy%'
+            or lower(pcr_patient_complaints) like '%bike%' or lower(pcr_patient_complaints) like '%bicy%' then 'Bicycle'
+        when
+            lower(pcr_cause_of_injury) like '%pede%'
+            or lower(pcr_patient_complaints) like '%pede%'
+            or (lower(pcr_cause_of_injury) like '%auto%' and lower(pcr_cause_of_injury) like '%ped%')
+            or (lower(pcr_patient_complaints) like '%auto%' and lower(pcr_patient_complaints) like '%ped%') then 'Pedestrian'
+        when
+            lower(pcr_cause_of_injury) like '%motorcy%'
+            or lower(pcr_patient_complaints) like '%mc %'
+            then 'Motorcycle'
+        when lower(mvc_form_vehicle_type) = 'motorized scooter' then 'E-Scooter'
+        else 'Motor Vehicle'
+    end
+) stored;
+
+
+--
+-- add some indexes
+--
+create index ems__incidents_crashes_crash_pk_index on public.ems__incidents (
+    crash_pk
+);
+create index ems__incidents_crashes_incident_received_datetime_index
+on public.ems__incidents (incident_received_datetime);
+create index ems__incidents_apd_incident_numbers_index on ems__incidents using gin (
+    apd_incident_numbers
+);
+create index ems__incidents_geometry_index on ems__incidents using gist (
+    geometry
+);
+create index ems__incidents_patient_injry_sev_id_index on ems__incidents (
+    patient_injry_sev_id
+);
+create index ems__incidents_crash_match_status_index on ems__incidents (
+    crash_match_status
+);
+create index crashes_case_id_index on crashes (case_id);
+create index crashes_is_temp_record_index on crashes (
+    is_temp_record
+);
+-- these indexes optimize our ST_DWithin operations, which require geography types and are the fastes way to check point-to-point proximity
+create index idx_ems_geometry_geography on ems__incidents using gist (
+    (geometry::geography)
+);
+create index idx_crashes_position_geography on crashes using gist (
+    (position::geography)
+);
+
+--
+-- Create or replace the trigger function for the crashes table
+--
+create or replace function update_crash_ems_match()
+returns trigger as $$
+DECLARE
+    matching_ems RECORD;
+    ems_record RECORD;
+    match_count INTEGER;
+    matched_crash_ids INTEGER[];
+    meters_threshold INTEGER := 1000;
+    time_threshold INTERVAL := '60 minutes';
+BEGIN
+    -- Find all EMS records near the crash location + time
+    FOR matching_ems IN (
+        SELECT 
+            e.id,
+            e.incident_received_datetime,
+            e.geometry
+        FROM 
+            ems__incidents e
+        WHERE
+            -- ignore manual qa matches: these should not be modified
+            e.crash_match_status != 'matched_by_manual_qa'
+            AND e.incident_received_datetime  >= (NEW.crash_timestamp - time_threshold)
+            AND e.incident_received_datetime  <= (NEW.crash_timestamp + time_threshold)
```

Comment:
i figured time threshold means we only want to link crashes and ems records that are within 60 min of each other, but this is implemented to allow a 2 hour difference. as is this would allow an ems record with `incident_received_datetime` of 2pm to be linked to a `crash_timestamp` of 4pm. just double checking this is whats intended

Reply:
how does it allow a crash thats 2 hours later / before to be linked?
incident received 2pm, crash_timestamp is 4pm
2pm isnt  >= 3pm, so even tho 2pm <= 5pm this wouldn't be linked, would it?

Reply:
in the example i was trying to do in my head i somehow decided that
4pm >= (2pm - 60min = 1pm) is true
4pm <= (2pm + 60min = 3pm) is also true

after looking at that for a very  long time ive decide 4pm is indeed not less than 3pm :melting_face: 

so nvm this is totally working as expected!

Reply:
idk, but as an latecomer to this PR and someone who is often confused about time, time window does help me think about the total time, 60 min, instead of in terms of +/- 30 min. ⏳ 🪟

Reply:
I ended up just adding a comment in the function args, because the dividing by a postgres `INTERVAL` type is not so simple.

This thread did remind me of a TIL from this work, which is that using `INTERVAL` for these selections is much faster than doing match on the epoch timestamps, which is how i first approached it:

```sql
and extract(epoch from e.incident_received_datetime) >= extract(epoch from NEW.crash_timestamp) - 1800
and extract(epoch from e.incident_received_datetime) <= extract(epoch from NEW.crash_timestamp) + 1800
```